File: Sources/WidgetBridge/WidgetRotation.swift
import Foundation

/// Deterministic hourly rotation, identical to the Kotlin `WidgetRotation`, so both platforms show the same item.
public enum WidgetRotation {
    private static let hourMs: Int64 = 3_600_000

    /// Local wall-clock hour slot; `zoneOffsetSeconds` is the zone offset at `epochMs`.
    public static func hourSlot(epochMs: Int64, zoneOffsetSeconds: Int) -> Int64 {
        floorDiv(epochMs + Int64(zoneOffsetSeconds) * 1_000, hourMs)
    }

    /// Convenience over `hourSlot(epochMs:zoneOffsetSeconds:)` for a `Date` in `zone`.
    public static func hourSlot(_ date: Date, in zone: TimeZone = .current) -> Int64 {
        hourSlot(epochMs: Int64((date.timeIntervalSince1970 * 1_000).rounded(.down)), zoneOffsetSeconds: zone.secondsFromGMT(for: date))
    }

    /// Index into a list of `count` items for `slot`; wraps and handles negative slots.
    public static func index(slot: Int64, count: Int) -> Int {
        precondition(count > 0, "count must be positive")
        let remainder = slot % Int64(count)
        return Int(remainder < 0 ? remainder + Int64(count) : remainder)
    }

    /// FNV-1a 32-bit over UTF-8 bytes, for stable per-item salts.
    public static func stableHash(_ value: String) -> UInt32 {
        var hash: UInt32 = 0x811C9DC5
        for byte in value.utf8 {
            hash = (hash ^ UInt32(byte)) &* 0x01000193
        }
        return hash
    }

    private static func floorDiv(_ dividend: Int64, _ divisor: Int64) -> Int64 {
        let quotient = dividend / divisor
        let remainder = dividend % divisor
        return (remainder != 0 && (remainder < 0) != (divisor < 0)) ? quotient - 1 : quotient
    }
}

File: widgetbridge/src/commonMain/kotlin/com/vocabloot/widgetbridge/WidgetRotation.kt
package com.vocabloot.widgetbridge

/** Deterministic hourly rotation shared by Kotlin and Swift readers so both platforms show the same item. */
public object WidgetRotation {
    private const val HOUR_MS = 3_600_000L

    /** Local wall-clock hour slot; [zoneOffsetSeconds] is the zone offset at [epochMs]. */
    public fun hourSlot(epochMs: Long, zoneOffsetSeconds: Int): Long = floorDiv(epochMs + zoneOffsetSeconds * 1_000L, HOUR_MS)

    /** Index into a list of [count] items for [slot]; wraps and handles negative slots. */
    public fun index(slot: Long, count: Int): Int {
        require(count > 0) { "count must be positive" }
        val remainder = slot % count.toLong()
        return (if (remainder < 0L) remainder + count else remainder).toInt()
    }

    /** FNV-1a 32-bit over UTF-8 bytes, for stable per-item salts. */
    public fun stableHash(value: String): UInt {
        var hash = 0x811C9DC5u
        value.encodeToByteArray().forEach { byte -> hash = (hash xor byte.toUByte().toUInt()) * 0x01000193u }
        return hash
    }

    private fun floorDiv(dividend: Long, divisor: Long): Long {
        val quotient = dividend / divisor
        val remainder = dividend % divisor
        return if (remainder != 0L && (remainder < 0L) != (divisor < 0L)) quotient - 1L else quotient
    }
}
